project_recalculate/models/project_project.py

```diff
@@ -0,0 +1,66 @@
+# See README.rst file on addon root folder for license details
+
+from odoo import models, fields, api, _
+from odoo.exceptions import UserError
+
+
+class ProjectProject(models.Model):
+    _inherit = 'project.project'
+
+    calculation_type = fields.Selection(
+        [('date_begin', 'Date begin'),
+         ('date_end', 'Date end')],
+        string='Calculation type', default=False,
+        help='How to calculate tasks, with date start or date end references. '
+             'If not set, "Recalculate project" button is disabled.')
+
+    def _start_end_dates_prepare(self):
+        """
+            Prepare project start or end date, looking into tasks list
+            and depending on project calculation_type
+            - if calculation_type == 'date_begin':
+                project end date = latest date from tasks end dates
+            - if calculation_type == 'date_end':
+                project start date = earliest date from tasks start dates
+
+            NOTE: Do not perform any write operations to DB
+        """
+        vals = {}
+        self.ensure_one()
+        if not self.tasks:
+            return vals
+        # Here we consider all project task, independently from the value in
+        # include_in_recalculate
+        start_task = min(self.tasks, key=lambda t: t.date_start or t.date_end)
```

Comment:
Shouldn't you filter out tasks without `include_in_recalculate`? :thinking:

Reply:
Well, as I've seen in the commit history, it has always been considered all project tasks, the ones in a stage with `include_in_recalculate = False` and the ones with `include_in_recalculate = True`. I do not know if I should change that.